### ISSUE TITLE
An internal error occurred while type checking file: RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by StockerMC** [microsoft/pyright#2438]
**Describe the bug**
When running `pyright discord/application_commands.py` in this [repository](https://github.com/StockerMC/discord.py), I get this error:
```
An internal error occurred while type checking file "C:\Users\user\Desktop\discord.py\discord\application_commands.py": RangeError: Maximum call stack size exceeded
    at Object.printType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typePrinter.ts:69:17)
    at Object.printType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:21309:28)
    at forEach (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typedDicts.ts:566:37) 
    at Map.forEach (<anonymous>)
    at canAssignTypedDict (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typedDicts.ts:560:17)
    at canAssignClass (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:17915:20)
    at canAssignType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:19248:22)
    at callback (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:19589:18)
    at C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeUtils.ts:213:13
    at Array.forEach (<anonymous>)
Error performing analysis: RangeError: Maximum call stack size exceeded
    at Object.printType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typePrinter.ts:69:17)
    at Object.printType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:21309:28)
    at forEach (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typedDicts.ts:566:37) 
    at Map.forEach (<anonymous>)
    at canAssignTypedDict (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typedDicts.ts:560:17)
    at canAssignClass (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:17915:20)
    at canAssignType (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:19248:22)
    at callback (C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeEvaluator.ts:19589:18)
    at C:\Users\user\AppData\Roaming\nvm\v16.9.1\node_modules\pyright\dist\pyright-internal\src\analyzer\typeUtils.ts:213:13
    at Array.forEach (<anonymous>)
```

**To Reproduce**
(Step 3 and 4 both give the same error)
1. `git clone https://github.com/StockerMC/discord.py`
2. `cd discord`
3. `pyright application_commands.py`
4. `pyright --lib`

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line (pyright 1.1.178)
This error also appears at the top of the same file with pylance version v2021.10.1

**Additional context**
I don't get this error on `pyright 1.1.177`. This seems to only be happening to the `application_commands.py` file. This also happens when using `pyright --lib` in the `discord` directory, however `pyright --verifytypes discord --ignoreexternal` doesn't give any error.

When removing the use of [`ApplicationCommandInteractionData`](https://github.com/StockerMC/discord.py/blob/0581b4f5055c52bae547b36e57b08720abbd79f3/discord/types/interactions.py#L165) (as a typehint) on [this line](https://github.com/StockerMC/discord.py/blob/0581b4f5055c52bae547b36e57b08720abbd79f3/discord/application_commands.py#L1093), the error goes away.
The error seems to be related to `TypedDict`s (as `typedDicts.ts` is in the stack trace).

**Comment by erictraut** [comment]
Thanks for reporting the problem. This was a long-standing bug that was exposed by a recent change. I've fixed the problem along with a few similar bugs in the code that could have caused a similar problem. The fix will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.179, which I just published. It will also be included in the next release of pylance.
